Introduce improved mechanism for util.Stopper.

spencerkimball opened the pull request:
In new conception, there is a draining phase. Any component which does
any unit of work as a "first mover" increments a drain count via
Stopper.Inc() and then decrements via Stopper.Dec() on completion. The
units of work which count are things like received external requests
and periodic work done in the gossip loop. While any of these things are
active, the stopper is prevented from initiating stoppage. However, all
first movers get false when calling Inc(), meaning they don't allow new
work to be scheduled.

The system quiesces when all draining is done at which point the stopper
signals all components to stop via ShouldStop() and they proceed to stop.

The idea behind the change is we thread an entire running system together
with a single stopper and are able to stop a complex system with more
confidence by allowing all activity to drain and then shutting down all
components.

The idea seems to have good results in terms of simplifying some of the
ugly stopping code which has been introduced of late. The gossip code got
a lot simpler as well.

Added a quitquitquit handler in the _admin REST API and corresponding CLI
command to quit a node. The draining we have isn't fully featured here,
but this is a start.

spencerkimball committed:
Introduce improved mechanism for util.Stopper.

In new conception, there is a draining phase. Any component which does
any unit of work as a "first mover" increments a drain count via
Stopper.Inc() and then decrements via Stopper.Dec() on completion. The
units of work which count are things like received external requests
and periodic work done in the gossip loop. While any of these things are
active, the stopper is prevented from initiating stoppage. However, all
first movers get false when calling Inc(), meaning they don't allow new
work to be scheduled.

The system quiesces when all draining is done at which point the stopper
signals all components to stop via ShouldStop() and they proceed to stop.

The idea behind the change is we thread an entire running system together
with a single stopper and are able to stop a complex system with more
confidence by allowing all activity to drain and then shutting down all
components.

The idea seems to have good results in terms of simplifying some of the
ugly stopping code which has been introduced of late. The gossip code got
a lot simpler as well.

Added a quit handler in the _admin REST API and corresponding CLI
command to quit a node. The draining we have isn't fully featured here,
but this is a start.

spencerkimball commented:
@tschottdorf here's what I've got so far. Doesn't all compile yet but I've gotten many of the unittests to run. It's a giant change, but also a big simplification.

petermattis commented on util/stopper.go:
```diff
@@ -17,39 +17,78 @@
+// all outstanding units of work have been completed via calls to
+// Dec(), then the stopper closes its stopper channel, which signals
+// components that it's safe to shut down. Once shutdown, each
+// component invokes SetStopped(). When all components has shutdown,
```
`s/has/have/g`

spencerkimball replied:
Done.

bdarnell commented on server/admin.go:
```diff
@@ -49,6 +50,8 @@ const (
 	// healthzPath is the healthz endpoint.
 	healthzPath = adminEndpoint + "healthz"
+	// quitPath is the quit endpoint.
+	quitPath = adminEndpoint + "quitquitquit"
```
We don't have to follow google's naming conventions here; everything is under `/_admin/` so we can keep the last component of the path simple without risk of collision. I'd just call this "quit" (and for that matter remove the z from "healthz").

Before we get to production status we'll also want to make any HTTP-accessible quit handler xsrf-protected instead of a simple get (and maybe just remove it completely or make it a development-only feature).

spencerkimball replied:
Changed to "quit" and "health"

bdarnell commented:
I like the new direction but the distinction between Add and Inc is subtle. How about introducing the names "task" and "worker":
* Add -> StartWorker
* SetStopped -> StopWorker
* ShouldStop -> ShouldStopWorkers (This would let us add a ShouldStopTasks channel if we need to signal that pending tasks may need to be aborted)
* Inc -> StartTask (or maybe TryStartTask to emphasize that it returns a bool that must be used, unlike StartWorker)
* Dec -> FinishTask

We could perhaps also add a Wait method that waits for the draining WaitGroup to reach zero without actually going into draining mode. This might give us a way to wait for the system to become quiescent without using IsTrueWithin so often.

spencerkimball commented:
I updated the naming in util.Stopper as expected. I also added a Quiesce() method, though it's not used anywhere yet. There was a unittest in storage/ for checking raft restarts that could use it, but we're not properly tracking all of the first mover work in raft/multiraft yet. I gave it a shot (as you'll see in the updated PR), but didn't get it. You should take a look, @bdarnell.

spencerkimball commented:
Fixes #608

spencerkimball commented:
Fixes #609

spencerkimball committed:
improved error logging and removed suspect code from RocksDB.Open()

spencerkimball commented:
@tschottdorf Not sure why the acceptance test is failing but it happens on my local machine too. This is an error from opening the RocksDB instance:

E0410 22:18:36.973407       6 start.go:155] Cockroach server exited with error: node.go:218: failed to start store: rocksdb.go:95: could not open RocksDB instance: IO error: lock /tmp/disk1/LOCK: No locks available

spencerkimball committed:
Tracked down cause of acceptance test failure.

The problem was we weren't including the engine as something to be
closed by the stopper in the case of bootstrapping a store.

spencerkimball commented:
@tschottdorf I'm having trouble ascertaining what circlci is complaining about in the latest failure.

tschottdorf commented:
@spencerkimball data race that didn't fail `make testrace`. There are many but I think they're all from `Stopper.Stop` vs `AddWorker` or something like that - so probably a missing Lock/atomic update in Stopper (haven't had time to look at it though). From `excerpt.txt` (or alternatively search for "DATA RACE" in `testrace.log`:

```
==================
WARNING: DATA RACE
Write by goroutine 36:
  sync.raceWrite()
      /usr/src/go/src/sync/race.go:41 +0x36
  sync.(*WaitGroup).Wait()
      /usr/src/go/src/sync/waitgroup.go:125 +0x177
  github.com/cockroachdb/cockroach/util.(*Stopper).Stop()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:104 +0xa8
  github.com/cockroachdb/cockroach/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/server/testserver.go:118 +0xa20
  github.com/cockroachdb/cockroach/client_test.StartTestServer()
      /go/src/github.com/cockroachdb/cockroach/client/client_test.go:49 +0x85
  github.com/cockroachdb/cockroach/client_test.TestKVClientRetryNonTxn()
      /go/src/github.com/cockroachdb/cockroach/client/client_test.go:103 +0x7a
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133

Previous read by goroutine 37:
  sync.raceRead()
      /usr/src/go/src/sync/race.go:37 +0x36
  sync.(*WaitGroup).Add()
      /usr/src/go/src/sync/waitgroup.go:63 +0xc1
  github.com/cockroachdb/cockroach/util.(*Stopper).AddWorker()
      /go/src/github.com/cockroachdb/cockroach/util/stopper.go:64 +0x4a
  github.com/cockroachdb/cockroach/storage.(*Range).startGossip()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:646 +0x43

Goroutine 36 (running) created at:
  testing.RunTests()
      /usr/src/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/src/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/cockroachdb/cockroach/client/_test/_testmain.go:110 +0x28c

Goroutine 37 (running) created at:
  github.com/cockroachdb/cockroach/storage.(*Range).start()
      /go/src/github.com/cockroachdb/cockroach/storage/range.go:243 +0xc0
  github.com/cockroachdb/cockroach/storage.(*Store).addRangeInternal()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:828 +0x88
  github.com/cockroachdb/cockroach/storage.func·018()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:428 +0x305
  github.com/cockroachdb/cockroach/storage/engine.MVCCIterate()
      /go/src/github.com/cockroachdb/cockroach/storage/engine/mvcc.go:993 +0x9be
  github.com/cockroachdb/cockroach/storage.(*Store).Start()
      /go/src/github.com/cockroachdb/cockroach/storage/store.go:442 +0xbb3
  github.com/cockroachdb/cockroach/server.BootstrapCluster()
      /go/src/github.com/cockroachdb/cockroach/server/node.go:135 +0x9ef
  github.com/cockroachdb/cockroach/server.(*TestServer).Start()
      /go/src/github.com/cockroachdb/cockroach/server/testserver.go:114 +0x8bc
  github.com/cockroachdb/cockroach/client_test.StartTestServer()
      /go/src/github.com/cockroachdb/cockroach/client/client_test.go:49 +0x85
  github.com/cockroachdb/cockroach/client_test.TestKVClientRetryNonTxn()
      /go/src/github.com/cockroachdb/cockroach/client/client_test.go:103 +0x7a
  testing.tRunner()
      /usr/src/go/src/testing/testing.go:447 +0x133
==================
```

tschottdorf commented:
@spencerkimball I've taken the liberty to add a commit that should hopefully address those data races (at least I didn't see any of them any more on my machine).
Feel free to do whatever you want with it.

tschottdorf commented:
added something else to handle shutting down the process from `runStart()`.

Haven't checked every nook and cranny in this PR, but it looks pretty good already and better what we have by far. We can figure out the rest later.

LGTM

petermattis commented on util/stopper.go:
```diff
@@ -17,39 +17,146 @@
-// Stopper instance.
+const (
+	// RUNNING is the state of a stopper before invoking Stop().
+	RUNNING = int32(0)
```
Any reason these constants are exported? And all-caps?

tschottdorf replied:
No. Forgot Go convention for a little while there. @spencerkimball I'll leave this further changes to you unless you tell me otherwise.

spencerkimball commented:
Thanks @tschottdorf. Looking at it now.

petermattis commented on util/stopper.go:
```diff
@@ -17,39 +17,146 @@
+// system is currently draining and the task should be refused.
+func (s *Stopper) StartTask() bool {
+	if s.getState() == RUNNING {
+		s.drain.Add(1)
```
I'm not seeing the point of the atomic operations for manipulating `Stopper.state`. And on the flip side it looks like it could be problematic. For example, between the `s.getState() == RUNNING` statement and the `s.drain.Add(1)` statement the stopper could close and finish.

spencerkimball replied:
This is no longer relevant. I removed the extra concurrency control around stopper. The trick was getting the calls to AddWorker to always happen before any chance of calling Stop.

petermattis commented on util/stopper.go:
```diff
@@ -17,39 +17,146 @@
+		return
+	}
+	s.setState(DRAINING)
+	s.drain.Wait()
```
Ah, here is spot where you wait for draining to clear while holding the lock. I think it would be better for `Stopper.drain` and `Stopper.stop` to be integers and to use a condition variable to wit for them to reach the desired values.

spencerkimball committed:
Revies Tobias' changes. Fix data race with WaitGroup by invoking AddWorker
correctly, in all cases before Stop() might ever be called in unittests.

spencerkimball committed:
Changed signature of callers to AddWorkers for clarity

bdarnell commented on gossip/gossip.go:
```diff
@@ -428,23 +424,15 @@ func (g *Gossip) manage() {
 			if g.outgoing.len()+g.incoming.len() == 0 {
 				log.Infof("no connections; signaling bootstrap")
-				g.stalled.Signal()
+				g.stalled <- struct{}{}
```
Is there any danger of the stalled channel reaching its 10-entry limit? If so, we should put this in a select to make the write non-blocking. We could also increase the limit here (why 10 instead of 1?); a high capacity on a `chan struct{}` is free.

bdarnell commented on gossip/server.go:
```diff
@@ -125,7 +125,10 @@ func (s *server) jitteredGossipInterval() time.Duration {
-func (s *server) start(rpcServer *rpc.Server) {
+func (s *server) start(rpcServer *rpc.Server, stopper *util.Stopper) {
+	stopper.AddWorker()
+	defer stopper.SetStopped()
```
Calling AddWorker and SetStopped in the same function is generally racy; we should move the call to AddWorker() to just before the `go s.start()` statement (it's important to be consistent here even if it's ok in this case, to make sure that the calls always match up).

spencerkimball replied:
How is it racy? This is entirely safe usage unless I'm missing something.

I tried doing it before the call and missed a bunch of cases in unittests which made me reconsider. Better to follow the same pattern everywhere which I've decided should be this:

    func (f *Foo) Start(stopper util.Stopper) {
      stopper.AddWorker()
      go func() {
        defer stopper.SetStopped()
        ...
      }()
    }

bdarnell replied:
We could even encapsulate this pattern in a `Stopper.RunWorker(f func())` method.

The race is that the goroutine may not get scheduled immediately, so this may stop too soon:
```
func f(stopper *util.Stopper) {
    stopper.AddWorker()
    defer stopper.SetStopped()
}

func g() {
  stopper := util.NewStopper()
  go f(stopper)
  stopper.Stop()
}
```

spencerkimball replied:
Good idea on the RunWorker method. Will update this PR.

bdarnell commented on gossip/client.go:
```diff
@@ -94,7 +94,9 @@ func (c *client) close() {
 // is modified and method returns for forwarding by caller.
-func (c *client) gossip(g *Gossip) error {
+func (c *client) gossip(g *Gossip, stopper *util.Stopper) error {
+	defer stopper.SetStopped()
```
Where is the AddWorker call that matches this SetStopped?

spencerkimball replied:
This has been moved into the start method.

bdarnell commented on multiraft/multiraft_test.go:
```diff
@@ -69,31 +73,21 @@ func newTestCluster(transport Transport, size int, t *testing.T) *testCluster {
-	}
-	for _, demux := range c.events {
-		demux.stop()
+		go node.start(stopper)
```
Doesn't each of these `go node.start` calls need to be paired with an AddWorker?

spencerkimball replied:
This is exactly why I changed the pattern to put all usage of the stopper by a struct into the same method (usually start).

bdarnell commented on server/cli/start.go:
```diff
@@ -145,7 +145,9 @@ func runStart(cmd *commander.Command, args []string) {
 	log.Info("Starting cockroach cluster")
-	s, err := server.NewServer(Context)
+	stopper := util.NewStopper()
+	stopper.AddWorker()
```
This AddWorker doesn't seem to be paired with a `go` statement or a `SetStopped`.

bdarnell replied:
Nevermind, I see the SetStopped below. This is an unusual pattern, though.

bdarnell commented on server/cli/start.go:
```diff
@@ -153,15 +155,55 @@ func runStart(cmd *commander.Command, args []string) {
 
 	c := make(chan os.Signal, 1)
-	signal.Notify(c, os.Interrupt, os.Kill)
+	signal.Notify(c, syscall.SIGINT, syscall.SIGKILL, syscall.SIGTERM)
```
Remove syscall.SIGKILL; it's not possible to handle that signal.

Does the move from `os` to `syscall` constants mean that won't even compile on windows? Maybe add a TODO to put this behind a build tag.

spencerkimball replied:
I've reverted to using os.Interrupt and os.Kill and added the SIGTERM below in a separate statement with a TODO.

bdarnell commented on server/cli/start.go:
```diff
@@ -153,15 +155,55 @@ func runStart(cmd *commander.Command, args []string) {
 
 	// Block until one of the signals above is received.
-	<-c
+	ch := make(chan struct{})
```
`c` and `ch` should have more distinct names.

spencerkimball replied:
Changed.

bdarnell commented on server/cli/start.go:
```diff
@@ -153,15 +155,55 @@ func runStart(cmd *commander.Command, args []string) {
+
+	select {
+	case <-c:
+		log.Warningf("SIGTERM or SIGKILL received, initiating hard shutdown")
```
s/SIGTERM or SIGKILL/second signal/

spencerkimball replied:
Done.

bdarnell commented on util/stopper.go:
```diff
@@ -17,39 +17,96 @@
+// have been completed via calls to FinishTask(), the stopper closes
+// its stopper channel, which signals all live workers that it's safe
+// to shut down. Once shutdown, each worker invokes SetStopped(). When
+// all workers have shutdown, the stopper is complete.
```
Add "closers" to this description.

spencerkimball replied:
Done.

bdarnell commented on util/stopper.go:
```diff
@@ -17,39 +17,96 @@
-	stopper chan struct{}
-	wg      sync.WaitGroup
+	stopper  chan struct{}
+	draining int32
```
Comment that `draining` uses atomic operations instead of `mu`.

spencerkimball replied:
I think you're looking at a pretty outdated diff now.

spencerkimball replied:
Strike that. OK, updated comment.

bdarnell commented on util/stopper.go:
```diff
@@ -17,39 +17,96 @@
+	closers  []Closer
 }
 
 // NewStopper returns an instance of Stopper. Count specifies how
```
The `count` argument is no longer present so you can remove the second sentence.

spencerkimball replied:
Done..

bdarnell commented:
LGTM

spencerkimball committed:
Addressed Ben's final feedback
